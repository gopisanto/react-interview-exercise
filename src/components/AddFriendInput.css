:local(.addFriendInput) {
  border-radius: 0;
  border-color: #ABAAAA;
  border-left: 0;
  border-right: 0;
}

:local(.gender) {
  background-color: white;
  border: 1px solid #ABAAAA;
  border-top: 0px;
  border-left: 0;
  border-right: 0;
  padding: 5px 0;
}

:local(.genderLabel) {
  margin: 0 10px;
}
.error {
  color: red;
}