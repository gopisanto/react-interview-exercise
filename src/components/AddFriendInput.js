import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { isEmpty } from 'lodash';
import styles from './AddFriendInput.css';

class AddFriendInput extends Component {

  render () {
    const {isDirty, errorMessages} = this.state;

    return (
      <div onKeyDown={this.handleSubmit.bind(this)}>
        {isDirty && errorMessages.length > 0 && <label className="error">Please provide {errorMessages.join(', ')}</label>}
        <input
          type="text"
          autoFocus="true"
          className={classnames('form-control', styles.addFriendInput)}
          placeholder="Type the name of a friend"
          value={this.state.name}
          onChange={this.handleChange.bind(this)} />
        <div className={styles.gender}>
          <label className={styles.genderLabel}>Gender:</label>
          <label className="radio-inline">
            <input
              type="radio"
              name="gender"
              value="male"
              onClick={this.handleGenderSelection}
              checked={this.state.gender === 'male'}
            />Male
          </label>
          <label className="radio-inline">
            <input
              type="radio"
              name="gender"
              value="female"
              onClick={this.handleGenderSelection}
              checked={this.state.gender === 'female'}
            />Female
          </label>
        </div>
      </div>
    );
  }

  constructor (props, context) {
    super(props, context);
    this.state = {
      name: this.props.name || '',
      gender: '',
      errorMessages: [],
      isDirty: false
    };
  }

  getErrorMessages = () => {
    const { name, gender } = this.state;
    const errorMessages = [];

    if (isEmpty(name)) {
      errorMessages.push('name');
    }
    if (isEmpty(gender)) {
      errorMessages.push('gender');
    }

    return errorMessages;
  }

  setErrorMessages = () => this.setState({errorMessages: this.getErrorMessages()});

  handleGenderSelection = event => {
    this.setState({gender: event.target.value}, this.setErrorMessages);
  }

  handleChange (e) {
    this.setState({name: e.target.value}, this.setErrorMessages);
  }

  handleSubmit (e) {
    const { name, gender } = this.state;
    const errorMessages = this.getErrorMessages();

    if (e.which === 13) {      
      if (errorMessages.length > 0) {
        this.setState({errorMessages, isDirty: true});
      } else {
        this.props.addFriend({ name: name.trim(), gender });
        this.setState({ name: '', gender: '', errorMessages: [], isDirty: false });
      }
    }
  }

}

AddFriendInput.propTypes = {
  addFriend: PropTypes.func.isRequired
};

export default AddFriendInput