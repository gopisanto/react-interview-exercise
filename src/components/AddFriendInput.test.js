import React from 'react';
import renderer from 'react-test-renderer';
import {configure, shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import sinon from 'sinon';
import AddFriendInput from './AddFriendInput';
import styles from './AddFriendInput.css';

configure({ adapter: new Adapter() });

describe('<AddFriendInput />', () => {
    it('should render', () => {
        const wrapper = renderer.create(<AddFriendInput addFriend={() => null} />).toJSON();

        expect(wrapper).toMatchSnapshot();
    });
    it('should say provide name, gender error when no data entered and submitted', () => {
        const wrapper = mount(<AddFriendInput addFriend={() => null} />);
        const errorMessage = 'Please provide name, gender';
        wrapper.instance().handleSubmit({which: 13});
        wrapper.update();
        expect(wrapper.find('.error').text()).toBe(errorMessage);
    });
    it('should say provide name error when no data entered and submitted', () => {
        const wrapper = mount(<AddFriendInput addFriend={() => null} />);
        const errorMessage = 'Please provide name';
        wrapper.setState({gender: 'male'});
        wrapper.instance().handleSubmit({which: 13});
        wrapper.update();
        expect(wrapper.find('.error').text()).toBe(errorMessage);
    });
    it('should say provide gender error when no data entered and submitted', () => {
        const wrapper = mount(<AddFriendInput addFriend={() => null} />);
        const errorMessage = 'Please provide gender';
        wrapper.setState({name: 'Santosh'});
        wrapper.instance().handleSubmit({which: 13});
        wrapper.update();
        expect(wrapper.find('.error').text()).toBe(errorMessage);
    });
    it('should call addFriend action on enter', () => {
        const spy = sinon.spy();
        const wrapper = mount(<AddFriendInput addFriend={spy} />);
        wrapper.setState({name: 'Santosh', gender: 'male'});
        wrapper.instance().handleSubmit({which: 13});
        expect(spy.calledOnce);
    });
});