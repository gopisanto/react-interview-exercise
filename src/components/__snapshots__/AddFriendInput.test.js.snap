exports[`<AddFriendInput /> should render 1`] = `
<div
  onKeyDown={[Function]}>
  <input
    autoFocus="true"
    className="form-control"
    onChange={[Function]}
    placeholder="Type the name of a friend"
    type="text"
    value="" />
  <div
    className={undefined}>
    <label
      className={undefined}>
      Gender:
    </label>
    <label
      className="radio-inline">
      <input
        checked={false}
        name="gender"
        onClick={[Function]}
        type="radio"
        value="male" />
      Male
    </label>
    <label
      className="radio-inline">
      <input
        checked={false}
        name="gender"
        onClick={[Function]}
        type="radio"
        value="female" />
      Female
    </label>
  </div>
</div>
`;
